features/26-mfa/api.md

```diff
@@ -0,0 +1,1003 @@
+function generateOTPAuthURIQRCodeImageURL(otpauthURI: string): string;
+
+// Register OOB
+
+type CreateNewOOBOptions = CreateNewOOBSMSOptions | CreateNewOOBEmailOptions;
+
+interface CreateNewOOBSMSOptions {
+  channel: "sms";
+  phone: string;
+}
+
+interface CreateNewOOBEmailOptions {
+  channel: "email";
+  email: string;
+}
+
+interface TriggerOOBOptions {
+  authenticatorID?: string;
+}
+
+interface ActivateOOBOptions {
+  authenticatorID: string;
+  code: string;
+}
+
+interface CreateNewOOBResult {
+  authenticatorID: string;
+  authenticatorType: "oob";
+  channel: "sms" | "email";
+}
+
+interface ActivateOOBResult {
+  recoveryCodes?: string[];
+}
+
+async function createNewOOB(options: CreateNewOOBOptions): Promise<CreateNewOOBResult>;
+async function triggerOOB(options?: TriggerOOBOptions): Promise<void>;
+async function activateOOB(opitons: ActivateOOBOptions): Promise<ActivateOOBResult>;
+
+// Error inspection
+
+interface AuthenticationSession {
+  token: string;
+  step: "identity" | "mfa";
+}
+
+interface AuthContainer {
+  authenticationSession?: AuthenticationSession;
+}
+
+// Authenticate
+
+interface AuthenticateWithTOTPOptions {
+  otp: string;
+  skipMFAForCurrentDevice?: boolean;
+}
+
+interface AuthenticateWithOOBOptions {
+  code: string;
+  skipMFAForCurrentDevice?: boolean;
+}
+
+interface AuthenticateWithRecoveryCodeOptions {
+  code: string;
+};
+
+async function authenticateWithTOTP(options: AuthenticateWithTOTPOptions): Promise<User>;
+
+async function authenticateWithOOB(options: AuthenticateWithOOBOptions): Promise<User>;
+
+async function authenticateWithRecoveryCode(options: AuthenticateWithRecoveryCodeOptions): Promise<User>;
+
+// Authenticator management
+
+type Authenticator = TOTPAuthenticator | OOBSMSAuthenticator | OOBEmailAuthenticator;
+
+interface TOTPAuthenticator {
+  id: string;
+  type: "totp";
+  activatedAt: Date;
+  displayName: string;
+}
+
+interface OOBSMSAuthenticator {
+  id: string;
+  type: "oob";
+  activatedAt: Date;
+  channel: "sms";
+  maskedPhone: string;
+}
+
+interface OOBEmailAuthenticator {
+  id: string;
+  type: "oob";
+  activatedAt: Date;
+  channel: "email";
+  maskedEmail: string;
+}
+
+interface RegenerateRecoveryCodeResult {
+  recoveryCodes: string[];
+}
+
+interface ListRecoveryCodeResult {
+  recoveryCodes: string[];
+}
+
+async function getAuthenticators(): Promise<Authenticator[]>;
+async function deleteAuthenticator(authenticatorID: string): Promise<void>;
+async function regenerateRecoveryCode(): Promise<RegenerateRecoveryCodeResult>;
+async function listRecoveryCode(): Promise<ListRecoveryCodeResult>;
+async function revokeAllTrustedDevices(): Promise<void>;
+```
+
+# Use Cases
+
+## Registration
+
+### SDK
+
+```typescript
+try {
+  await skygear.auth.login("user@example.com", "password");
+} catch (e) {
+  const authnSession = skygear.auth.authenticationSession;
+  if (authnSession != null && authnSession.step === "mfa") {
```

Comment:
As comments on the old version, the `catch` without using authentication feel weird.

And would love to see if we can eventually remove the whole `authenticationSession` concept from users

Reply:
The reason why I added `authenticationSession` is that we want to support some application that may reload during the authentication session.

If authentication session is only stored in memory and the application reloads, the authentication session is clear. Effectively the user can never login.

If we choose not to support such use case, the authentication session can be hidden from the developer. We can use `e` and a helper function to tell the developer the step.